src/metabase/models/query/permissions.clj

```diff
@@ -196,38 +231,37 @@
       (let [paths-excluding-gtap-paths (set/difference paths (-> gtap-perms :paths))]
         (or (perms/set-has-full-permissions-for-set? @api/*current-user-permissions-set* paths-excluding-gtap-paths)
             (throw (perms-exception paths)))))
-    ;; Check native query access if required
-    (when (= (:perms/create-queries required-perms) :query-builder-and-native)
-      (or (= (:perms/create-queries gtap-perms) :query-builder-and-native)
-          (= (data-perms/full-db-permission-for-user api/*current-user-id* :perms/create-queries db-id) :query-builder-and-native)
-          (throw (perms-exception {db-id {:perms/create-queries :query-builder-and-native}}))))
-    (when (= (:perms/view-data required-perms) :unrestricted)
-      (or (= (:perms/view-data gtap-perms) :unrestricted)
-          (= :unrestricted (data-perms/full-db-permission-for-user api/*current-user-id* :perms/view-data db-id))
-          (throw (perms-exception {db-id {:perms/view-data :unrestricted}}))))
-    (when-let [table-id->perm (and (coll? (:perms/create-queries required-perms))
-                                   (:perms/create-queries required-perms))]
-      (doseq [[table-id _] table-id->perm]
-        (or
-         (contains? #{:query-builder :query-builder-and-native}
-                    (get-in gtap-perms [:perms/create-queries table-id]))
-         (data-perms/user-has-permission-for-table?
-          api/*current-user-id*
-          :perms/create-queries
-          :query-builder
-          db-id
-          table-id)
-         (throw (perms-exception {db-id {:perms/create-queries {table-id :query-builder}}})))))
+
+    ;; Check view-data and create-queries permissions, for individual tables or the entire DB
+    (doseq [perm-type [:perms/view-data :perms/create-queries]]
```

Comment:
I've refactored this code to reduce duplication and enforce `required-perms` more comprehensively. It'll now enforce any combination of `required-perms` passed into this function, so we don't have to worry about things breaking if `required-perms-for-query` changes.